From 8d5186eed0ee67fee96e0fc87a30485bf551b551 Mon Sep 17 00:00:00 2001
From: "Leonidas S. Barbosa" <leo.barbosa@canonical>
Date: Mon, 6 Apr 2020 18:00:14 -0300
Subject: [PATCH] Fix-test-bug79282

---
 ext/exif/tests/bug79282.phpt | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/ext/exif/tests/bug79282.phpt b/ext/exif/tests/bug79282.phpt
index 7b7e3656..afeda730 100644
--- a/ext/exif/tests/bug79282.phpt
+++ b/ext/exif/tests/bug79282.phpt
@@ -7,9 +7,9 @@ var_dump(exif_read_data('data://image/jpeg;base64,/9jhAAlFeGlmAAAg'));
 
 ?>
 --EXPECTF--
-Warning: exif_read_data(): Invalid TIFF alignment marker in %s on line %d
+Warning: exif_read_data(9jhAAlFeGlmAAAg): Missing TIFF alignment marker in %s on line %d
 
-Warning: exif_read_data(): File structure corrupted in %s on line %d
+Warning: exif_read_data(9jhAAlFeGlmAAAg): File structure corrupted in %s on line %d
 
-Warning: exif_read_data(): Invalid JPEG file in %s on line %d
+Warning: exif_read_data(9jhAAlFeGlmAAAg): Invalid JPEG file in %s on line %d
 bool(false)
-- 
2.25.1

